pkg/agent/flexvolume/manager/ceph/manager.go

```diff
@@ -0,0 +1,180 @@
+	}
+
+	// attach and poll until volume is mapped
+	logger.Infof("attaching volume %s/%s cluster %s", pool, image, clusterName)
+	monitors, keyring, err := getClusterInfo(vm.context, clusterName)
+	defer os.Remove(keyring)
+	if err != nil {
+		return "", fmt.Errorf("failed to load cluster information from cluster %s: %+v", clusterName, err)
+	}
+
+	err = cephclient.MapImage(vm.context, image, pool, clusterName, keyring, monitors)
+	if err != nil {
+		return "", fmt.Errorf("failed to map image %s/%s cluster %s. %+v", pool, image, clusterName, err)
+	}
+
+	// poll for device path
+	retryCount := 0
+	maxRetries := 10
+	for {
+		devicePath, err := vm.devicePathFinder.FindDevicePath(image, pool, clusterName)
+		if err != nil {
+			return "", fmt.Errorf("failed to poll for mapped image %s/%s cluster %s. %+v", pool, image, clusterName, err)
+		}
+
+		if devicePath != "" {
+			return devicePath, nil
+		}
+
+		retryCount++
+		if retryCount >= maxRetries {
+			return "", fmt.Errorf("exceeded retry count while finding device path: %+v", err)
+		}
+
+		logger.Infof("failed to find device path, sleeping 1 second: %+v", err)
+		<-time.After(time.Second)
+	}
+}
+
+// Detach the volume
+func (vm *VolumeManager) Detach(image, pool, clusterName string) error {
+	// check if the volume is attached
+	devicePath, err := vm.isAttached(image, pool, clusterName)
+	if err != nil {
+		return fmt.Errorf("failed to check if volume %s/%s is attached cluster %s", pool, image, clusterName)
+	}
+	if devicePath == "" {
+		logger.Infof("volume %s/%s is already detached cluster %s", pool, image, clusterName)
+		return nil
+	}
+
+	logger.Infof("detaching volume %s/%s cluster %s", pool, image, clusterName)
+	monitors, keyring, err := getClusterInfo(vm.context, clusterName)
+	defer os.Remove(keyring)
+	if err != nil {
+		return fmt.Errorf("failed to load cluster information from cluster %s: %+v", clusterName, err)
+	}
+
+	err = cephclient.UnMapImage(vm.context, image, pool, clusterName, keyring, monitors)
+	if err != nil {
+		return fmt.Errorf("failed to detach volume %s/%s cluster %s. %+v", pool, image, clusterName, err)
+	}
+	logger.Infof("detached volume %s/%s", pool, image)
+	return nil
+}
+
+// Check if the volume is attached
+func (vm *VolumeManager) isAttached(image, pool, clusterName string) (string, error) {
+	devicePath, err := vm.devicePathFinder.FindDevicePath(image, pool, clusterName)
+	if err != nil {
+		return "", err
+	}
+	return devicePath, nil
+}
+
+func getClusterInfo(context *clusterd.Context, clusterName string) (string, string, error) {
+	clusterInfo, _, err := mon.LoadClusterInfo(context, clusterName)
+	if err != nil {
+		return "", "", fmt.Errorf("failed to load cluster information from cluster %s: %+v", clusterName, err)
+	}
+
+	// create temp keyring file
+	keyringFile, err := ioutil.TempFile("", clusterName+".keyring")
+	if err != nil {
+		return "", "", err
+	}
+
+	keyring := fmt.Sprintf(cephmon.AdminKeyringTemplate, clusterInfo.AdminSecret)
```

Comment:
Yeah that is because the `rbd map` command takes a filename for the keyring or keyfile. http://docs.ceph.com/docs/jewel/man/8/rbd/

They used to take the secret directly but that has been deprecated.